### ISSUE TITLE
fromfile add a header dtype

### ISSUE THREAD

**Issue by mcspritz** [numpy/numpy#11195]
Hi,

I (unfortunately) use LabView and found that it saves binary files including the number of data written at the beginning of the file.

Is it possible to add a parameter like a header dtype to extract this value or  to skip them?

Thanks

**Comment by pxcandeias** [comment]
Hi Luca,

disclaimer: I am not part of the numpy development team.

I do not think this is the appropriate forum to discuss this issue but, since I too have to deal with binary files created by LabVIEW, I suggest that you create a function to read those files tailored to your needs (that is what I did). You can still use numpy.frofile but with multiple calls, using the 'count' argument to read first the size of the array and then the full array from the remainder of the file (eventually checking the size to see if it matches), supposing that only one array is stored per file. The file read position is memorised between calls so you will get consecutive data chunks between calls.

That way you can also deal with the binary type of the array (through the 'dtype' argument), which might be different from the header part (int32 according to the documentation available on their website), and not forgetting byte endianness.

Best regards,
Paulo

**Comment by pxcandeias** [comment]
Hi Luca,

it is precisely because of the (possible) mixture of data types in the file to be read that I recommend to write a specialized function. Otherwise, how can you instruct numpy.fromfile() to convert the binary data? Its purpose is to read an ndarray of a single data type, not mixed types.

From the documentation of numpy.fromfile():
> dtype : data-type
>> Data type of the returned array. For binary files, it is used to determine the size and byte-order of the items in the file.

Best regards,
Paulo

**Comment by mcspritz** [comment]
But dtype can be constructed as mixed data types
from [numpy.dtype](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.dtype.html)
`np.dtype({'names':['gender','age'], 'formats':['S1',np.uint8]})
dtype([('gender', '|S1'), ('age', '|u1')])
`

**Comment by pxcandeias** [comment]
Hi Luca,

you are right about that. Let us wait then to hear someone from the numpy development team give their opinion.

Best regards,
Paulo

**Comment by charris** [comment]
IIRC, SciPy has some specialized imports. @rgommers @pv Thoughts?

**Comment by rgommers** [comment]
The only somewhat interesting use of `fromfile` in SciPy is in `scipy/io/_fortran.py`. Not sure if that's of any help here.

**Comment by charris** [comment]
@rgommers I was thinking of `scipy/io` that supports various special files types and thought it might be the place for a LabView addition.

**Comment by rgommers** [comment]
Ah okay. Maybe - if we're happy with a reader for the proprietary IDL, then LabView can at least be considered as well. 

@mcspritz @pxcandeias is this a common, documented and stable file format? I haven't touched LabView in 10 years.

**Comment by charris** [comment]
Maybe related, IIRC, Enthought was doing some work with LabView.

**Comment by mattip** [comment]
The suggestion to create a wrapper function to handle LabView-specific data and then continue with the same file pointer to call `np.fromfile` seems like the most productive path forward for you.

**Comment by pxcandeias** [comment]
Hello everybody,

I have developed what I now consider a "production ready" module to deal with LabVIEW flattened data strings that I would gladly share with the rest of the world, that is the reason why I created https://github.com/pxcandeias/py-LFDS in the first place, but I have not done it yet because:
1) this would be the first time I would do such a thing and I am not at ease with the whole process;
2) I do have little spare time to solve any bugs that might be detected.

It is completely based on numpy and its dtypes to perform the conversion from the binary LabVIEW data types, taking into account byte endianness.
If you think that could be interesting, I could use some assistance to make it finally public or even to be incorporated somehow in scipy, although I do not know how that could be done.

Best regards,
Paulo

**Comment by pxcandeias** [comment]
Hi @rgommers,

the LabVIEW flattened data strings are documented at

https://zone.ni.com/reference/en-XX/help/371361P-01/lvconcepts/flattened_data/

Best regards,
Paulo

**Comment by rgommers** [comment]
Thanks @pxcandeias. That seems like reasonable documentation for the format. It would be good to have it published as a separate module (like https://github.com/pxcandeias/py-LFDS) before considering adding it to SciPy. Once it has seen some use and positive user feedback, we could reconsider.

> If you think that could be of interest, I could use some assistance/guidance to make it finally public or even to incorporate it somehow in scipy, although I do not know how that could be done.

@mcspritz maybe you would be interested in collaborating? @pxcandeias it may be useful to just put your implementation in https://github.com/pxcandeias/py-LFDS (just as a Python file that you have already), and then ask for help of other people interested in LabView on the numpy-discussion or one of the scipy mailing lists.

**Comment by mcspritz** [comment]
Hi all,

To be honest I'm not a good fun of LabView, we are using a four years old version and the "Write to Binary File" function ([here](http://zone.ni.com/reference/en-XX/help/371361P-01/glang/write_file/)) is still the same since.

Does it have to be LabView specific? I mean, if you decide to implement the header it could be applied to other cases as well, I did not find any at the moment but I'm sure it could be valuable for somebody else.

I will look into the py-LFDS git.

Thanks

**Comment by pv** [comment]
Regarding the part of this discussion relevant for numpy: There's no
need to have "header" support in fromfile. It is a low-level routine
for reading from files that reads from current file position.

You can call it twice or seek the file position to correct place to
skip headers.

**Comment by mcspritz** [comment]
Thanks all.

I find this a little absurd because you call a low-level routine but then let use complex dtypes?
Can somebody explain be why a low-level routine can have high level complex dtypes?

from the documentation you cannot "call it twice or seek the file position to skip headers". Or someone has to rewrite the documentation to include this case.

**Comment by eric-wieser** [comment]
> you cannot "call it twice or seek the file position to skip headers"

Does the following not work?

```python
with open('thefile', 'rb') as f:
    header = f.read(4)
    data = np.fromfile(f, dtype=[('a', np.uint32), ('b', np.int16'), ('b', np.uint32')])
```

**Comment by pxcandeias** [comment]
Hi everybody,

it took me while to set up everything in GitHub (remember I said it was my first try?) but I finally managed to have a decent summary for the module I have been developing. During this process I renamed it to LV_binary_files (it now resides at https://github.com/pxcandeias/LV_binary_files).

As I try to explain there, the LVBF.py module defines a class, named LV_fd(), which is intended to be used at a low level to read/write flattened data. It should be called through a higher level class which sets the .fobj attribute as the handle for an open file object that is subsequently used by numpy.fromfile (reading) and numpy.tofile (writing).

I hope this helps @mcspritz, and anyone else, in dealing with LabVIEW flattened data binary files.
